File: zflap-core/src/pda.rs
use std::collections::HashSet;

use serde::Serialize;

#[derive(Clone, Debug)]
pub struct PdaTransition {
    pub from: String,
    pub to: String,
    pub input: Option<char>,  // None = epsilon
    pub pop: Option<char>,    // None = no pop required
    pub push: String,         // "" = push nothing (epsilon)
}

#[derive(Clone, Debug, Serialize)]
pub struct PdaStep {
    pub from_state: String,
    pub to_state: String,
    pub consumed: Option<char>,
    pub popped: Option<char>,
    pub pushed: String,
    pub stack_snapshot: String,
    pub input_index: usize,
}

#[derive(Clone, Debug)]
struct Config {
    state: String,
    input_index: usize,
    stack: Vec<char>,
}

#[derive(Clone, Debug)]
pub struct Pda {
    initial_state: String,
    pub initial_stack_symbol: char,
    final_states: HashSet<String>,
    transitions: Vec<PdaTransition>,
}

impl Pda {
    pub fn new(initial_state: String, initial_stack_symbol: char) -> Self {
        Self {
            initial_state,
            initial_stack_symbol,
            final_states: HashSet::new(),
            transitions: Vec::new(),
        }
    }

    pub fn add_transition(&mut self, t: PdaTransition) {
        self.transitions.push(t);
    }

    pub fn add_final_state(&mut self, state: String) {
        self.final_states.insert(state);
    }

    pub fn remove_final_state(&mut self, state: &str) {
        self.final_states.remove(state);
    }

    /// Returns (accepted, execution path). Uses DFS with backtracking.
    pub fn accepts(&self, input: &str, max_steps: usize) -> (bool, Vec<PdaStep>) {
        let chars: Vec<char> = input.chars().collect();
        let config = Config {
            state: self.initial_state.clone(),
            input_index: 0,
            stack: vec![self.initial_stack_symbol],
        };
        let mut path = Vec::new();
        let mut steps = 0usize;
        let found = self.dfs(&config, &chars, &mut path, max_steps, &mut steps);
        (found, path)
    }

    fn dfs(
        &self,
        config: &Config,
        input: &[char],
        path: &mut Vec<PdaStep>,
        max_steps: usize,
        steps: &mut usize,
    ) -> bool {
        if *steps >= max_steps {
            return false;
        }
        *steps += 1;

        if config.input_index == input.len() && self.final_states.contains(&config.state) {
            return true;
        }

        let top = config.stack.last().copied();

        for trans in &self.transitions {
            if trans.from != config.state {
                continue;
            }

            let input_ok = match trans.input {
                None => true,
                Some(c) => config.input_index < input.len() && input[config.input_index] == c,
            };
            if !input_ok {
                continue;
            }

            let pop_ok = match trans.pop {
                None => true,
                Some(c) => top == Some(c),
            };
            if !pop_ok {
                continue;
            }

            let mut new_stack = config.stack.clone();
            if trans.pop.is_some() {
                new_stack.pop();
            }
            // Push chars in reverse so the first char ends up on top
            for c in trans.push.chars().rev() {
                new_stack.push(c);
            }

            let new_index = if trans.input.is_some() {
                config.input_index + 1
            } else {
                config.input_index
            };

            path.push(PdaStep {
                from_state: config.state.clone(),
                to_state: trans.to.clone(),
                consumed: trans.input,
                popped: trans.pop,
                pushed: trans.push.clone(),
                stack_snapshot: stack_str(&new_stack),
                input_index: new_index,
            });

            let next = Config {
                state: trans.to.clone(),
                input_index: new_index,
                stack: new_stack,
            };

            if self.dfs(&next, input, path, max_steps, steps) {
                return true;
            }

            path.pop();
        }

        false
    }

    pub fn get_transitions(&self) -> &[PdaTransition] {
        &self.transitions
    }
}

fn stack_str(stack: &[char]) -> String {
    stack.iter().rev().collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Builds a PDA for a^n b^n (n >= 1).
    fn anbn_pda() -> Pda {
        // q0: push A for each 'a'
        // q1: pop A for each 'b'
        // q2: final
        // Stack alphabet: Z (bottom), A
        let mut pda = Pda::new("q0".into(), 'Z');
        pda.add_final_state("q2".into());

        // q0 on 'a', top=Z: push AZ (A on top of Z)
        pda.add_transition(PdaTransition {
            from: "q0".into(), to: "q0".into(),
            input: Some('a'), pop: Some('Z'), push: "AZ".into(),
        });
        // q0 on 'a', top=A: push AA
        pda.add_transition(PdaTransition {
            from: "q0".into(), to: "q0".into(),
            input: Some('a'), pop: Some('A'), push: "AA".into(),
        });
        // q0 -> q1 on 'b', top=A: pop A
        pda.add_transition(PdaTransition {
            from: "q0".into(), to: "q1".into(),
            input: Some('b'), pop: Some('A'), push: "".into(),
        });
        // q1 on 'b', top=A: pop A
        pda.add_transition(PdaTransition {
            from: "q1".into(), to: "q1".into(),
            input: Some('b'), pop: Some('A'), push: "".into(),
        });
        // q1 on epsilon, top=Z: go to q2
        pda.add_transition(PdaTransition {
            from: "q1".into(), to: "q2".into(),
            input: None, pop: Some('Z'), push: "Z".into(),
        });

        pda
    }

    #[test]
    fn test_anbn_accepts() {
        let pda = anbn_pda();
        assert!(pda.accepts("ab", 10_000).0);
        assert!(pda.accepts("aabb", 10_000).0);
        assert!(pda.accepts("aaabbb", 10_000).0);
    }

    #[test]
    fn test_anbn_rejects() {
        let pda = anbn_pda();
        assert!(!pda.accepts("aab", 10_000).0);
        assert!(!pda.accepts("abb", 10_000).0);
        assert!(!pda.accepts("", 10_000).0);
        assert!(!pda.accepts("ba", 10_000).0);
    }
}

File: .claude/src/core/PDA.cpp
#include "PDA.h"
#include <algorithm>

using namespace std;

PDA::PDA(const std::string &initialState, char initialStackSymbol)
    : initialState(initialState), initialStackSymbol(initialStackSymbol) {}

void PDA::addTransition(const PDA_Transition &t) {
    transitions.push_back(t);
}

void PDA::addFinalState(const std::string &s) {
    finalStates.insert(s);
}

string PDA::stackToString(const stack<char> &s) {
    stack<char> tmp = s;
    string out;
    while (!tmp.empty()) {
        out.push_back(tmp.top());
        tmp.pop();
    }
    return out; // top..bottom
}

optional<PDA_Step> PDA::getStepFromPath(const vector<PDA_Step> &path, size_t i) const {
    if (i >= path.size()) return nullopt;
    return path[i];
}

bool PDA::accepts(const std::string &input, std::vector<PDA_Step> *outPath, int maxSteps) {
    Config start;
    start.state = initialState;
    start.inputIndex = 0;
    start.stack = std::stack<char>();
    start.stack.push(initialStackSymbol);

    vector<PDA_Step> pathSoFar;
    vector<PDA_Step> resultPath;
    int stepsRemaining = maxSteps;

    bool found = dfs_find(input, start, pathSoFar, resultPath, stepsRemaining);
    if (found && outPath) *outPath = resultPath;
    return found;
}

bool PDA::dfs_find(const std::string &input,
                   Config current,
                   vector<PDA_Step> &pathSoFar,
                   vector<PDA_Step> &resultPath,
                   int &stepsRemaining) {
    if (stepsRemaining-- <= 0) return false; // guard against infinite epsilon cycles

    // Accept if all input consumed and in a final state.
    if ((int)current.inputIndex == (int)input.size() && finalStates.count(current.state)) {
        resultPath = pathSoFar;
        return true;
    }

    for (const auto &t : transitions) {
        if (t.from != current.state) continue;

        // Check whether the input symbol matches (or is epsilon).
        bool inputMatches = (t.input == '\0') ||
                            (current.inputIndex < (int)input.size() &&
                             input[current.inputIndex] == t.input);
        if (!inputMatches) continue;

        // Check and perform pop.
        std::stack<char> newStack = current.stack;
        char popped = '\0';
        if (t.pop != '\0') {
            if (newStack.empty() || newStack.top() != t.pop) continue;
            popped = newStack.top();
            newStack.pop();
        }

        // Perform push: the first character of push ends up deepest,
        // so we push in reverse order.
        for (auto it = t.push.rbegin(); it != t.push.rend(); ++it) {
            newStack.push(*it);
        }

        PDA_Step step;
        step.fromState   = current.state;
        step.toState     = t.to;
        step.consumed    = t.input;
        step.popped      = popped;
        step.pushed      = t.push;
        step.inputIndex  = current.inputIndex + (t.input == '\0' ? 0 : 1);
        step.stackSnapshot = stackToString(newStack);

        Config next;
        next.state      = t.to;
        next.inputIndex = step.inputIndex;
        next.stack      = newStack;

        pathSoFar.push_back(step);
        if (dfs_find(input, next, pathSoFar, resultPath, stepsRemaining)) return true;
        pathSoFar.pop_back();
    }

    return false;
}
